eth/fetcher/block_fetcher.go

```diff
@@ -596,16 +595,14 @@ func (f *BlockFetcher) loop() {
 			blocks := []*types.Block{}
 			// abort early if there's nothing explicitly requested
 			if len(f.completing) > 0 {
-				for i := 0; i < len(task.blockHashes) && i < len(task.transactions) && i < len(task.randomness) && i < len(task.epochSnarkData); i++ {
+				// for i := 0; i < len(task.blockHashes) && i < len(task.transactions) && i < len(task.randomness) && i < len(task.epochSnarkData); i++ {
+				for i := 0; i < len(task.transactions) && i < len(task.randomness) && i < len(task.epochSnarkData); i++ {
 					// Match up a body to any possible completion request
 					var matched = false
 					for hash, announce := range f.completing {
 						if f.queued[hash] != nil || announce.origin != task.peer {
 							continue
 						}
-						if task.blockHashes[i] != announce.header.Hash() {
-							continue
-						}
```

Comment:
@mcortesi @gastonponti This seems to be the fundamental change that passing the hashes allows for. Do you see any reason for that?

Reply:
Looks like this was a [part](https://github.com/celo-org/celo-blockchain/pull/952) of a [bug](https://github.com/celo-org/celo-blockchain/issues/950) fix. I'm worrying that authors haven't added a test that tests this situation, reverting it looks risky, no?

Reply:
Sure, but this is something we need to fix anyways when we want to separate protocols.

The only explanation that I have is that we could mistake an uncle block for  the correct one a take the wrong chain?
```go
if task.blockHashes[i] != announce.header.Hash() {
	continue
}
```